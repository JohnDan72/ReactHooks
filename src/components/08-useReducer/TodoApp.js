import React, { useEffect, useReducer } from 'react';
import JSONTree from 'react-json-tree';
import { Col, Divider, FlexboxGrid } from 'rsuite';

// hooks
import { todoReducer } from './todoReducer';

import FormTask from './FormTask';
import ListTodos from './ListTodos';

// const initialState = [
//     {
//         id: new Date().getTime(),
//         desc: 'Aprender React',
//         done: false
//     }
// ];

const init = () => {
    return JSON.parse(localStorage.getItem('state')) || [];
}

const TodoApp = () => {

    const [state, dispatch] = useReducer(todoReducer, [], init);
    

    useEffect(() => {
        localStorage.setItem('state', JSON.stringify(state));
    }, [state]);

    const handleDelete = (idItem) => {
        const action = {
            type: 'D_ITEM',
            payload: idItem
        }

        dispatch(action);
    }

    const handleTogle = (idItem) => {
        dispatch({
            type: 'TOGGLE',
            payload: idItem
        });
    }

    const handleAdd = (newTodo) => {
        dispatch({
            type: 'C_ITEM',
            payload: newTodo
        });
    }
    

    return (
        <>
            <h3>TodoApp (total: {state.length})</h3>
            <Divider></Divider>

            <FlexboxGrid align="middle">
                <FlexboxGrid.Item as={Col} xs={24} sm={12}>
                    
                    <FormTask   handleAdd={handleAdd}/>

                    <ListTodos  state={state}
                                handleTogle={handleTogle}
                                handleDelete={handleDelete}
                                />


                </FlexboxGrid.Item>
                <FlexboxGrid.Item as={Col} xs={24} sm={12}>
                    <Divider>state</Divider>
                    <JSONTree data={state} />

                </FlexboxGrid.Item>
            </FlexboxGrid>


        </>
    );
}


export default TodoApp;