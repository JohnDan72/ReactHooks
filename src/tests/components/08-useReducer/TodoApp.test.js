const { shallow } = require("enzyme")
import TodoApp from "../../../components/08-useReducer/TodoApp";

describe('Pruebas TodoApp', () => {


    let wrapper = shallow(<TodoApp />);

    test('debe renderizar el component', () => {
        // console.log(wrapper.props())
        expect(wrapper).toMatchSnapshot();
    })
    
})
