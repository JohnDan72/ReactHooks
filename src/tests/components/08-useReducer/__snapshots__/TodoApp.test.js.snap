// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas TodoApp debe renderizar el component 1`] = `
<Fragment>
  <h3>
    TodoApp (total: 
    0
    )
  </h3>
  <Divider />
  <FlexboxGrid
    align="middle"
  >
    <FlexboxGridItem
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "as": [Function],
            "className": [Function],
            "classPrefix": [Function],
            "lg": [Function],
            "lgHidden": [Function],
            "lgOffset": [Function],
            "lgPull": [Function],
            "lgPush": [Function],
            "md": [Function],
            "mdHidden": [Function],
            "mdOffset": [Function],
            "mdPull": [Function],
            "mdPush": [Function],
            "sm": [Function],
            "smHidden": [Function],
            "smOffset": [Function],
            "smPull": [Function],
            "smPush": [Function],
            "xs": [Function],
            "xsHidden": [Function],
            "xsOffset": [Function],
            "xsPull": [Function],
            "xsPush": [Function],
          },
          "render": [Function],
        }
      }
      sm={12}
      xs={24}
    >
      <FormTask
        handleAdd={[Function]}
      />
      <ListTodos
        handleDelete={[Function]}
        handleTogle={[Function]}
        state={Array []}
      />
    </FlexboxGridItem>
    <FlexboxGridItem
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "as": [Function],
            "className": [Function],
            "classPrefix": [Function],
            "lg": [Function],
            "lgHidden": [Function],
            "lgOffset": [Function],
            "lgPull": [Function],
            "lgPush": [Function],
            "md": [Function],
            "mdHidden": [Function],
            "mdOffset": [Function],
            "mdPull": [Function],
            "mdPush": [Function],
            "sm": [Function],
            "smHidden": [Function],
            "smOffset": [Function],
            "smPull": [Function],
            "smPush": [Function],
            "xs": [Function],
            "xsHidden": [Function],
            "xsOffset": [Function],
            "xsPull": [Function],
            "xsPush": [Function],
          },
          "render": [Function],
        }
      }
      sm={12}
      xs={24}
    >
      <Divider>
        state
      </Divider>
      <JSONTree
        collectionLimit={50}
        data={Array []}
        getItemString={[Function]}
        hideRoot={false}
        invertTheme={true}
        isCustomNode={[Function]}
        keyPath={
          Array [
            "root",
          ]
        }
        labelRenderer={[Function]}
        postprocessValue={[Function]}
        shouldExpandNode={[Function]}
        valueRenderer={[Function]}
      />
    </FlexboxGridItem>
  </FlexboxGrid>
</Fragment>
`;
